Experimental Hard Arrange

JackyP opened the pull request:
A somewhat hacky implementation for sort inspired by terasort - use hard_group_by, with semi-sorted .out.disk.frame.ids, and then chunk_arranging the results. 

It implements arrange(x, y) and arrange(desc(x), y) but no other expressions in the arrange, and there are R deprecation warnings.

This is my first time creating a pull request on an open source package, and I'm still wrapping my head around the package internals which was my motivation for putting this together, but I thought I'd show you for inspiration.

Once again great work with the package!

xiaodaigh commented:
Thank you very much @JackyP I will be reviewing it soon.

xiaodaigh commented on tests/testthat/test-hard-arrange.R:
```diff
@@ -0,0 +1,41 @@
+  dff = csv_to_disk.frame("tmp_pls_delete_gb.csv", "tmp_pls_delete_gb.df")
+  
+  # Sort ascending
+  sorted_dff <- dff %>% hard_arrange(id1, id4)
```
I am getting warning

```r
Warning message:
Using `as.character()` on a quosure is deprecated as of rlang 0.3.0.
Please use `as_label()` or `as_name()` instead.
```

We need to get rid of all warnings before submissions to CRAN.

JackyP replied:
Rewrote that section to use rlang::quo_is_symbol. This removes the deprecation warning.

xiaodaigh commented on R/hard_arrange.r:
```diff
@@ -0,0 +1,60 @@
+  # Limitations: this does not evaluate expressions - e.g. arrange by f(x)
+  q <- quos(...)
+  
+  vars <- sapply(q, function(x){ 
```
Keen to get an URL to the exact way dplyr handles `desc`.

JackyP replied:
https://github.com/tidyverse/dplyr/blob/master/src/arrange.cpp

I'm thinking that assigning the expressions that are not symbols to temporary variables may be the way to go.

xiaodaigh commented on R/hard_arrange.r:
```diff
@@ -0,0 +1,60 @@
+
+#' @rdname hard_arrange
+#' @importFrom purrr map
+#' @importFrom purrr map
```
duplicated with previous line

JackyP replied:
Removed

xiaodaigh commented on R/sortablestr2i.R:
```diff
@@ -0,0 +1,42 @@
+             )
+  }
+  
+  do.call(
```
add comments about what this is doing and examples of output. I think we should separate this out as a function for generating the `if-else` and have some tests cases.

xiaodaigh replied:
I mean the `switchcond` function and the below

JackyP replied:
Function separated, comments added. Not sure how to best set up the test cases.

xiaodaigh commented on tests/testthat/test-hard-arrange.R:
```diff
@@ -0,0 +1,41 @@
+  # Sort ascending
+  sorted_dff <- df %>% hard_arrange(id1, id4)
+  sorted_df <- sorted_dff %>% collect
+  expect_true(!is.unsorted(sorted_df$id1))
```
need more tests to test that the second sort by variables is well sorted

JackyP replied:
Tests updated.

xiaodaigh replied:
The test might be hard to interpret. I was thinking of a more straight-forward test, you can group by the first level and then for each subdataframe check for sortedness?

xiaodaigh commented on tests/testthat/test-hard-arrange.R:
```diff
@@ -0,0 +1,41 @@
+  desc_dff <- df %>% hard_arrange(desc(id4), id2)
+  desc_df <- desc_dff %>% collect
+  expect_true(!is.unsorted(-desc_df$id4))
+})
```
add a third test that that tests `hard_arrange` with three by variables.

JackyP replied:
Test updated.

xiaodaigh commented on tests/testthat/test-hard-arrange.R:
```diff
@@ -0,0 +1,41 @@
+setup({
+  
+  df = disk.frame:::gen_datatable_synthetic(1e3+11)
+  data.table::fwrite(df, "tmp_pls_delete_gb.csv")
```
I have moved all of these files to `tempdir()` in the `development` branch. You may want to merge with `development` branch and do a PR to `development` branch

JackyP replied:
Updated PR to development branch.

xiaodaigh commented on R/sortablestr2i.R:
```diff
@@ -0,0 +1,42 @@
+# shard_by_rule <- splitstr2i(split_values)
+# code = glue::glue("df[,.out.disk.frame.id := {shard_by_rule}]")
+
+sortablestr2i <- function(split_values, desc_vars){
```
I am a little concerned about the performance of this. I am not sure if it will slow, but it seems the fastest possible in R. I know `{data.table}` has `fifelse` but not sure how efficient that is for multi-statement `if`s.

Can we have some benchmarks that can be run? The `hash2stri` can compute billions of hashes every quickly.

JackyP replied:
This was surprisingly not that much slower in my testing. 

Try this which benchmarks on 30x flights dataset (10m records):
 * hard_group_by process with "hash" vs the new "sort" splitter
 * the proposed disk.frame arrange vs dplyr arrange, and
 * the hash function vs the generated sort split function

```{r}
library(dplyr)
library(data.table)
library(disk.frame)
n_rep = 30
flights.df <- rbindlist(lapply(seq_len(n_rep), function(X) nycflights13::flights))
flights.dff <- as.disk.frame(flights.df)
flights.dt <- as.data.table(flights.df)

# Time to hard group by with hash:
tic <- Sys.time()
hgb1 <- flights.dff %>%
  hard_group_by(origin, year, month, day, shardby_function="hash")
(toc = Sys.time() - tic)
  
# Time to hard group by with sort:  
tic <- Sys.time()
hgb2 <- flights.dff %>%
  hard_group_by(origin, year, month, day, shardby_function="sort")
(toc = Sys.time() - tic)
  
# Time to sort with disk.frame:  
tic <- Sys.time()
harr <- flights.dff %>%
  hard_arrange(origin, year, month, day)
(toc = Sys.time() - tic)
  
# Time to Dplyr sort
tic <- Sys.time()
darr <- flights.df %>%
  arrange(origin, year, month, day)
(toc = Sys.time() - tic)

# Raw test of the functions - hashstr2i
hashstr2i <- function(x, ngrps) {
  .Call('_disk_frame_hashstr2i', PACKAGE = 'disk.frame', x, ngrps)
}

tic <- Sys.time()
flights.df[,.out.disk.frame.id := hashstr2i(paste0(day,month,origin,year), 4)]
(toc = Sys.time() - tic)

# Raw test of the functions - sortablestr2i
tic <- Sys.time()
flights.dt[,.out.disk.frame.id := ((origin > "EWR" | (origin == "EWR" & (year > 2013 | (year == 2013 & (month > 4 | (month == 4 & (day > 26 | (day == 26)))))))))* 1 + ((origin > "EWR" | (origin == "EWR" & (year > 2013 | (year == 2013 & (month > 9 | (month == 9 & (day > 19 | (day == 19)))))))))* 1 + ((origin > "JFK" | (origin == "JFK" & (year > 2013 | (year == 2013 & (month > 2 | (month == 2 & (day > 19 | (day == 19)))))))))* 1 + ((origin > "JFK" | (origin == "JFK" & (year > 2013 | (year == 2013 & (month > 7 | (month == 7 & (day > 17 | (day == 17)))))))))* 1 + ((origin > "JFK" | (origin == "JFK" & (year > 2013 | (year == 2013 & (month > 11 | (month == 11 & (day > 19 | (day == 19)))))))))* 1 + ((origin > "LGA" | (origin == "LGA" & (year > 2013 | (year == 2013 & (month > 4 | (month == 4 & (day > 4 | (day == 4)))))))))* 1 + ((origin > "LGA" | (origin == "LGA" & (year > 2013 | (year == 2013 & (month > 8 | (month == 8 & (day > 26 | (day == 26)))))))))* 1]
(toc = Sys.time() - tic)
```
Noting that I am on an older non-SSD Mac:
1. The "sort split" function is only about 50% slower than the hash2stri function.
2. This contributes a 10% slower time for hard_group_by overall
3. Overall this disk.frame hard_arrange is about four times slower than dplyr arrange. 

YMMV with an SSD.

xiaodaigh commented on R/hard_group_by.r:
```diff
@@ -108,10 +109,18 @@ hard_group_by.data.frame <- function(df, ..., add = FALSE, .drop = FALSE) {
+  stopifnot(shardby_function %in% c("hash", "sort"))
+  
+  if (shardby_function == "sort" && is.null(sort_splits)){
+    sort_splits <- map(df, sample_n, size=ceiling(nchunks / disk.frame::nchunks(df))) %>% 
```
can we sample more? I think sampling `nchunks` rows from each `chunk`, collect them and then pick out the cut points from the `nchunks^2` rows would be better.

Also, what if there are `duplicated` from each sample? I think it will still work, but we may want to generate some warnings in some cases.

JackyP replied:
Added more sampling - default 100 per. Open to suggestions on how to generate warnings!

xiaodaigh commented on R/hard_group_by.r:
```diff
@@ -108,10 +109,41 @@ hard_group_by.data.frame <- function(df, ..., add = FALSE, .drop = FALSE) {
+    # Get splits. May lead to less than nchunks if duplicates are selected.
+    sort_splits <- sort_splits_sample %>% 
+      slice(ntiles) %>%
+      distinct()
```
should it be `distinct` and then `slice`? Say we wanted 100 chunks, but only 88 distinct cuts. Then we should give a warning at that point.

JackyP replied:
I thought it should be slice then distinct. Let's say you have 100 records in sort_splits_sample, you want to have four chunks at the end.

We want slices to be asymptotically representative of the 25th, 50th and 75th percentile ranks (by sort order) in the original dataset. 

So the logic here is to get record numbers 25, 50, 75 as the cuts. 

If we have one very common combination, then the 50th and 75th percentile could be the same, in which case we end up with one very big chunk, and only three chunks. That's where I'm thinking we need to put in a warning. And update nchunks to be 3 instead of 4 - need to find where I do that.

On the other hand, if we distinct first, then:
1. the sample weights are no longer representative of the weights they represent in the big dataset. E.g. we could have one group that has a lot of records in the big dataset, but they only count as one in the sample
2. we could end up with 88 records left, so row 75 will be 75/88 rather than 75/100

xiaodaigh replied:
Just one last note! I will fix it. Thanks for this awesome PR!

You just need to add `dplyr::sample_n` and `dplyr::slice` in place of `sample_n` and `slice`. Notes won't pass CRAN either.

```r
N  checking R code for possible problems (7.2s)
   hard_group_by.disk.frame: no visible binding for global variable
     'sample_n'
   hard_group_by.disk.frame: no visible global function definition for
     'slice'
   Undefined global functions or variables:
     sample_n slice
```

JackyP commented:
No pain - thanks for reviewing this. 
Also need to add a test for datetimes, the logic works for numeric and text but not sure about dates...

JackyP commented:
>  group by the first level and then for each subdataframe check 

With the hard-to-interpret tests, I'm worried about the performance of that and how ugly the logic might get if we recurse to the three levels of groupings. How about if we sort the reference dataframe using dplyr and compare the columns for differences?

xiaodaigh commented:
Also seems to be failing on the iris example in the hard_arrange since the generate query doesn't quote the strings.

JackyP commented:
> checking Rd \usage sections ... WARNING
Undocumented arguments in documentation object 'hard_group_by'
  'sort_splits' 'desc_vars' 'sort_split_sample_size'

> Undocumented arguments in documentation object 'rechunk'
  'shardby_function' 'sort_splits' 'desc_vars'

> Undocumented arguments in documentation object 'shard'
  'desc_vars'

> Undocumented arguments in documentation object 'write_disk.frame'
  'shardby_function' 'sort_splits' 'desc_vars'

With the new parameters, is your preference to expose them, or wrap the functions and hide them for a cleaner API?

xiaodaigh commented:
I think the way you have done is fine. Those error message are from the checking process. We need eliminate all warnings. I think we just need more @param to eliminate them.

JackyP commented:
> I am not passing the local checks on my Windows machine.
> 
> ```r
> * checking Rd \usage sections ... WARNING
> Undocumented arguments in documentation object 'hard_group_by'
>   'sort_splits' 'desc_vars' 'sort_split_sample_size'
> 
> Undocumented arguments in documentation object 'rechunk'
>   'shardby_function' 'sort_splits' 'desc_vars'
> 
> Undocumented arguments in documentation object 'shard'
>   'desc_vars'
> 
> Undocumented arguments in documentation object 'write_disk.frame'
>   'shardby_function' 'sort_splits' 'desc_vars'
> ```
> 
> Sorry to be a pain, but any warnings in the checks will not pass CRAN either. We need to make sure there is 0 warnings and 0 notes, when do the check `--as-cran`.
> 
> Also _this is important_, when you run the checks, please rename the `test` folder to `test_manual`, or the checks will never pass. I don't submit the tests folder to CRAN, cos there are weird errors that I can't get pass Checks.

So far I've added logic and tests for handling factors (causing the iris failure) and dates. I have also updated the previously undocumented variables in docs. 

However, I am unable to run checks. It says that 

```
E  creating vignettes (18s)
   Quitting from lines 55-62 (glm.Rmd) 
   Error: processing vignette 'glm.Rmd' failed with diagnostics:
   no applicable method for 'bigglm' applied to an object of class "function"
   Execution halted
```

I have biglm_0.9-1 which I only just installed from CRAN. I'm not familiar with that package, any ideas why it might be failing? The help file does seem to say that formulas are fine.

xiaodaigh commented:
Don't have internet at home atm. But try to remove the cache in the
vignette and if it still fails, just rename the vignette folder to
something else when doing a check.

On Mon., 14 Oct. 2019, 22:20 JackyP, <notifications@github.com> wrote:

> I am not passing the local checks on my Windows machine.
>
> * checking Rd \usage sections ... WARNINGUndocumented arguments in documentation object 'hard_group_by'
>   'sort_splits' 'desc_vars' 'sort_split_sample_size'
> Undocumented arguments in documentation object 'rechunk'
>   'shardby_function' 'sort_splits' 'desc_vars'
> Undocumented arguments in documentation object 'shard'
>   'desc_vars'
> Undocumented arguments in documentation object 'write_disk.frame'
>   'shardby_function' 'sort_splits' 'desc_vars'
>
> Sorry to be a pain, but any warnings in the checks will not pass CRAN
> either. We need to make sure there is 0 warnings and 0 notes, when do the
> check --as-cran.
>
> Also *this is important*, when you run the checks, please rename the test
> folder to test_manual, or the checks will never pass. I don't submit the
> tests folder to CRAN, cos there are weird errors that I can't get pass
> Checks.
>
> So far I've added logic and tests for handling factors (causing the iris
> failure) and dates. I have also updated the previously undocumented
> variables in docs.
>
> However, I am unable to run checks. It says that
>
> E  creating vignettes (18s)
>    Quitting from lines 55-62 (glm.Rmd)
>    Error: processing vignette 'glm.Rmd' failed with diagnostics:
>    no applicable method for 'bigglm' applied to an object of class "function"
>    Execution halted
>
> I have biglm_0.9-1 which I only just installed from CRAN. I'm not familiar
> with that package, any ideas why it might be failing? The help file does
> seem to say that formulas are fine.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/xiaodaigh/disk.frame/pull/189?email_source=notifications&email_token=ABCJ6JJW7FWMYTB4VX7K5XDQORIXFA5CNFSM4I7NM6KKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBEHRDA#issuecomment-541620364>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABCJ6JMOU7DBSRR6RIQFVUTQORIXFANCNFSM4I7NM6KA>
> .
>

JackyP commented:
I've updated the tests per our earlier discussion of comparing to dplyr. 

The remaining glm problem with the checks was with the make_glm_streaming_fn. Removing the verbose parameter from the inner function resolved the issue. See latest commit.

I am expecting checks to pass now, please let me know how it goes when you have the internet back!
